choisohyun/2020-11-26-leave-site-vue.md

```diff
@@ -0,0 +1,155 @@
+# Vue에서 'Leave Site?'를 구현하는 3가지 방법
+
+>> 웹사이트에서 내용을 입력하다가 창을 벗어나려고 하면 `이 사이트에서 나가시겠습니까?`와 같은 경고창을 본 적이 있을 것입니다.
+이번 포스팅에서는 Vue를 사용하여 사이트를 벗어나지 못하도록 막는 방법 3가지를 간단한 예제 사이트를 만들면서 알아보겠습니다.
+
+## 개요
+
+1. 예제 사이트 구성
+2. 다른 사이트로 나가는 것 감지하기 - `beforeunload` 이벤트
+3. 다른 라우터 주소로 이동하는 것 감지하기 - vue-router의 `beforeRouteLeave` 
+4. 다른 쿼리스트링 주소로 이동하는 것 감지하기 - `custom method`
+
+## 1. 예제 사이트 구성
+
+> 다양한 상황에서 경고창을 띄우기 위해 라우터를 나누어 구성하였습니다.
+
+path: repository와 detail-issue로 설정했습니다. `localhost:8080/repository`와 같은 주소에서 설정한 컴포넌트를 볼 수 있습니다.
+
+```jsx
+routes: [
+  {
+    path: "/repository",
+    component: Repository
+  },
+  {
+    path: "/detail-issue",
+    component: DetailIssue
+  }
+]
+```
+
+query: 탭을 3가지 만들어 vue-router에서 지원하는 router-link로 query를 만들어 줍니다. query는 `/repository?tabName={type}`과 같은 주소로 보여집니다.
+
+```html
+<router-link :to="{ path: $route.path, query: { tabName: type } }" @click="checkCanLeavePage()">
+  {{ name }}
+</router-link>
+```
+
+아래와 같은 라우터로 구성되어 있습니다. Input 태그가 있는 CreateIssue 컴포넌트가 사이트 나가기를 결정지어야 하는 곳입니다. 값이 입력되면 사이트를 바로 나갈 수 없게 막아야 하기 때문입니다.
+
+![leave-site-1](./images/leave-site.jpeg)
+
+Git Repository 링크: [https://github.com/choisohyun/leave-site-example](https://github.com/choisohyun/leave-site-example)
+
+## 2. 다른 사이트로 나가는 것 감지하기 - `beforeunload` 이벤트
+
+가장 먼저 쉽게 할 수 있는 방법입니다. window에서 지원하는 beforeunload 이벤트인데요. document가 언로드되기 전에 호출되는 이벤트입니다. 
+
+beforeunload 이벤트는 구체적으로 다음과 같은 상황에서 발생합니다.
+
+- 사이트 창을 닫으려고 할 때
+- 다른 주소로 이동하려고 할 때
+- 같은 주소여도 새로고침하려고 할 때(이 때는 `Reload site?` 문구가 뜹니다)
+
+![leave-site-2](./images/leave-site-2.png)
+
+이제 코드를 작성해 봅시다. 
+
+여기서 Input을 감지하기 위해 v-model과 watch로 값의 변화를 감지하고 `canLeaveSite`라는 boolean 값으로 결정하도록 합니다. 
+
+그리고 vue의 라이프 사이클 주기 중에서 mounted될 때와 beforeUnmount될 때 이벤트를 등록하고 해지시키도록 합니다. 
+
+```jsx
+	mounted() {
+    window.addEventListener('beforeunload', this.unLoadEvent);
+  },
+  beforeUnmount() {
+    window.removeEventListener('beforeunload', this.unLoadEvent);
+  },
```

Comment:
```suggestion
```jsx
  mounted() {
    window.addEventListener('beforeunload', this.unLoadEvent);
  },
  beforeUnmount() {
    window.removeEventListener('beforeunload', this.unLoadEvent);
  },
```